crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,345 @@
+                })
+                .collect::<Result<HashMap<String, RegisterMatrix>, RegisterMatrixConversionError>>(
+                )?,
+        ))
     }
 
-    /// `readout_values` maps program-defined readout to result-defined readout, e.g.:
-    ///     { "ro[0]": "q0", "ro[1]": "q1" }
-    /// where `ro[0]` is defined in the original program, and `q0` is what comes back in execution results.
-    /// Here we map the result-defined readout values back to their original result-defined names.
-    pub(crate) fn from_mappings_and_values(
-        readout_mappings: &HashMap<String, String>,
-        readout_values: &HashMap<String, ReadoutValues>,
-    ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
+    /// Attempts to build a [`ReadoutMap`] from [`QpuReadout`].
+    ///
+    /// # Errors
+    ///
+    /// This fails if the underlying [`QpuReadout`] data is jagged. See [`ReadoutData`] for more
+    /// detailed explanations of why and when this occurs.
+    pub fn from_qpu_readout_data(
+        readout_data: &QpuReadout,
+    ) -> Result<Self, RegisterMatrixConversionError> {
+        Ok(
+            Self(
+                readout_data
+                    .mappings
                     .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
-
-        result
-    }
-}
+                    // Pair all the memory references with their readout values
+                    .map(|(memory_reference, alias)| {
+                        Ok((
+                            parse_readout_register(memory_reference).map_err(|e| {
+                                RegisterMatrixConversionError::MemoryReferenceParseError(e)
+                            })?,
+                            readout_data.readout_values.get(alias).ok_or(
+                                RegisterMatrixConversionError::UnmappedAlias {
+                                    memory_reference: memory_reference.to_string(),
+                                    alias: alias.to_string(),
+                                },
+                            )?,
+                        ))
+                    })
+                    // Collect into a type that will sort them by memory reference, this allows us
+                    // to make sure indices are sequential.
+                    .collect::<Result<
+                        BTreeMap<MemoryReference, &ReadoutValues>,
+                        RegisterMatrixConversionError,
+                    >>()?
+                    .iter()
```

Comment:
Also useful to avoid unnecessary cloning later

```suggestion
                    .into_iter()
```